File: src/main/java/Vista/ProveedorPanel.java
package Vista;

import javax.swing.*;
import javax.swing.table.DefaultTableModel;
import Controlador.*;
import Modelo.*;
import java.awt.*;
import java.util.List;

public class ProveedorPanel extends JPanel {

	private final ProveedorControlador ctrl;
	private final Usuario usuario;
	private JTabbedPane tabs;
	private DefaultTableModel modeloProveedores;
	private DefaultTableModel modeloOrdenes;
	private DefaultTableModel modeloCuentas;

	public ProveedorPanel(ProveedorControlador ctrl, Usuario usuario) {
		this.ctrl = ctrl;
		this.usuario = usuario;
		setLayout(new BorderLayout(0, 16));
		setBackground(Estilos.BG_CLARO);
		setBorder(BorderFactory.createEmptyBorder(24, 24, 24, 24));
		construir();
	}

	private void construir() {
		JPanel header = new JPanel(new BorderLayout());
		header.setBackground(Estilos.BG_BLANCO);
		header.setBorder(BorderFactory.createCompoundBorder(BorderFactory.createLineBorder(Estilos.BORDE),
				BorderFactory.createEmptyBorder(20, 24, 20, 24)));
		JLabel tit = new JLabel("ABASTECIMIENTO  —  Proveedores y Cuentas por Pagar");
		tit.setFont(Estilos.FUENTE_TITULO);
		header.add(tit, BorderLayout.WEST);

		tabs = new JTabbedPane();
		tabs.setFont(Estilos.FUENTE_BOLD);
		tabs.addTab("Directorio", construirTabProveedores());
		tabs.addTab("Ordenes", construirTabOrdenes());
		tabs.addTab("Deudas", construirTabCuentas());

		add(header, BorderLayout.NORTH);
		add(tabs, BorderLayout.CENTER);
	}

	private JPanel construirTabProveedores() {
		JPanel p = new JPanel(new BorderLayout(0, 8));
		p.setBackground(Estilos.BG_BLANCO);
		p.setBorder(BorderFactory.createEmptyBorder(12, 12, 12, 12));

		JPanel toolbar = new JPanel(new FlowLayout(FlowLayout.RIGHT));
		toolbar.setBackground(Estilos.BG_BLANCO);
		JButton btnNuevo = Estilos.botonPrimario("+ Alta Proveedor");
		btnNuevo.setPreferredSize(new Dimension(160, 40));
		btnNuevo.addActionListener(e -> abrirFormularioProveedor(null));
		toolbar.add(btnNuevo);

		String[] cols = { "ID", "Empresa", "Contacto", "Telefono", "Email", "Direccion" };
		modeloProveedores = new DefaultTableModel(cols, 0) {
			@Override
			public boolean isCellEditable(int r, int c) {
				return false;
			}
		};
		JTable tabla = new JTable(modeloProveedores);
		tabla.setFont(Estilos.FUENTE_NORMAL);
		tabla.setRowHeight(40);
		tabla.setShowGrid(false);
		tabla.getTableHeader().setFont(Estilos.FUENTE_XS);
		tabla.getTableHeader().setBackground(Estilos.BG_ZINC_100);
		tabla.getColumnModel().getColumn(0).setMaxWidth(60);

		tabla.addMouseListener(new java.awt.event.MouseAdapter() {
			@Override
			public void mouseClicked(java.awt.event.MouseEvent e) {
				if (e.getClickCount() == 2) {
					int row = tabla.getSelectedRow();
					if (row >= 0) {
						int id = (int) modeloProveedores.getValueAt(row, 0);
						Proveedor prov = ctrl.getProveedores().stream().filter(x -> x.getId() == id).findFirst()
								.orElse(null);
						if (prov != null)
							abrirFormularioProveedor(prov);
					}
				}
			}
		});

		JScrollPane scroll = new JScrollPane(tabla);
		scroll.setBorder(BorderFactory.createLineBorder(Estilos.BORDE));
		p.add(toolbar, BorderLayout.NORTH);
		p.add(scroll, BorderLayout.CENTER);
		return p;
	}

	private JPanel construirTabOrdenes() {
		JPanel p = new JPanel(new BorderLayout(0, 8));
		p.setBackground(Estilos.BG_BLANCO);
		p.setBorder(BorderFactory.createEmptyBorder(12, 12, 12, 12));

		JPanel toolbar = new JPanel(new FlowLayout(FlowLayout.RIGHT));
		toolbar.setBackground(Estilos.BG_BLANCO);
		JButton btnNueva = Estilos.botonExito("+ Generar Pedido");
		btnNueva.setPreferredSize(new Dimension(170, 40));
		btnNueva.addActionListener(e -> abrirNuevaOrden());
		toolbar.add(btnNueva);

		String[] cols = { "Folio", "Proveedor", "Total", "Pago", "Estado", "Accion" };
		modeloOrdenes = new DefaultTableModel(cols, 0) {
			@Override
			public boolean isCellEditable(int r, int c) {
				return c == 5;
			}
		};
		JTable tablaOrdenes = new JTable(modeloOrdenes);
		tablaOrdenes.setFont(Estilos.FUENTE_NORMAL);
		tablaOrdenes.setRowHeight(42);
		tablaOrdenes.setShowGrid(false);
		tablaOrdenes.getTableHeader().setFont(Estilos.FUENTE_XS);
		tablaOrdenes.getTableHeader().setBackground(Estilos.BG_ZINC_100);
		tablaOrdenes.getColumnModel().getColumn(0).setMaxWidth(100);
		tablaOrdenes.getColumnModel().getColumn(3).setMaxWidth(100);
		tablaOrdenes.getColumnModel().getColumn(4).setMaxWidth(110);
		tablaOrdenes.getColumnModel().getColumn(5).setMaxWidth(110);

		tablaOrdenes.getColumnModel().getColumn(5)
				.setCellRenderer(new VentaPanel.BtnRenderer("Recibir", Estilos.EMERALD_CLARO, Estilos.EMERALD));
		tablaOrdenes.getColumnModel().getColumn(5).setCellEditor(new VentaPanel.BtnEditor(id -> {
			try {
				int oid = Integer.parseInt(id);
				OrdenCompra orden = ctrl.getOrdenes().stream().filter(o -> o.getId() == oid).findFirst().orElse(null);
				if (orden != null && orden.getEstado() == OrdenCompra.Estado.RECIBIDO) {
					JOptionPane.showMessageDialog(SwingUtilities.getWindowAncestor(this), "Esta orden ya fue recibida.",
							"Aviso", JOptionPane.INFORMATION_MESSAGE);
					return;
				}
				ctrl.recibirOrden(oid);
				refrescarOrdenes();
				refrescarCuentas();
			} catch (Exception ignored) {
			}
		}));

		JScrollPane scroll = new JScrollPane(tablaOrdenes);
		scroll.setBorder(BorderFactory.createLineBorder(Estilos.BORDE));
		p.add(toolbar, BorderLayout.NORTH);
		p.add(scroll, BorderLayout.CENTER);
		return p;
	}

	private JPanel construirTabCuentas() {
		JPanel p = new JPanel(new BorderLayout(0, 8));
		p.setBackground(Estilos.BG_BLANCO);
		p.setBorder(BorderFactory.createEmptyBorder(12, 12, 12, 12));

		JLabel subtit = new JLabel("  Saldos pendientes con acreedores");
		subtit.setFont(Estilos.FUENTE_SMALL);
		subtit.setForeground(Estilos.TEXTO_TENUE);

		String[] cols = { "Proveedor", "Orden", "Total", "Pagado", "Saldo", "Estado", "Accion" };
		modeloCuentas = new DefaultTableModel(cols, 0) {
			@Override
			public boolean isCellEditable(int r, int c) {
				return c == 6;
			}
		};
		JTable tablaCuentas = new JTable(modeloCuentas);
		tablaCuentas.setFont(Estilos.FUENTE_NORMAL);
		tablaCuentas.setRowHeight(42);
		tablaCuentas.setShowGrid(false);
		tablaCuentas.getTableHeader().setFont(Estilos.FUENTE_XS);
		tablaCuentas.getTableHeader().setBackground(Estilos.BG_ZINC_100);
		tablaCuentas.getColumnModel().getColumn(5).setMaxWidth(100);
		tablaCuentas.getColumnModel().getColumn(6).setMaxWidth(120);

		tablaCuentas.getColumnModel().getColumn(6)
				.setCellRenderer(new VentaPanel.BtnRenderer("Liquidar", Estilos.INDIGO_CLARO, Estilos.INDIGO));
		tablaCuentas.getColumnModel().getColumn(6).setCellEditor(new VentaPanel.BtnEditor(id -> {
			try {
				int cid = Integer.parseInt(id);
				CuentaPorPagar cuenta = ctrl.getCuentas().stream().filter(c -> c.getId() == cid).findFirst()
						.orElse(null);
				if (cuenta != null && cuenta.getSaldo() > 0) {
					String input = JOptionPane.showInputDialog(SwingUtilities.getWindowAncestor(this),
							"Monto a liquidar (saldo: $" + String.format("%.2f", cuenta.getSaldo()) + "):",
							String.format("%.2f", cuenta.getSaldo()));
					if (input != null) {
						double monto = Double.parseDouble(input.trim());
						ctrl.liquidarCuenta(cid, monto);
					}
				}
			} catch (Exception ignored) {
			}
		}));

		JScrollPane scroll = new JScrollPane(tablaCuentas);
		scroll.setBorder(BorderFactory.createLineBorder(Estilos.BORDE));
		p.add(subtit, BorderLayout.NORTH);
		p.add(scroll, BorderLayout.CENTER);
		return p;
	}

	public void refrescarProveedores() {
		if (modeloProveedores == null)
			return;
		modeloProveedores.setRowCount(0);
		for (Proveedor p : ctrl.getProveedores())
			modeloProveedores.addRow(new Object[] { p.getId(), p.getNombre(), p.getContacto(), p.getTelefono(),
					p.getEmail(), p.getDireccion() });
	}

	public void refrescarOrdenes() {
		if (modeloOrdenes == null)
			return;
		modeloOrdenes.setRowCount(0);
		for (OrdenCompra o : ctrl.getOrdenes()) {
			if (o.getEstado() == OrdenCompra.Estado.PENDIENTE) {
				modeloOrdenes.addRow(new Object[] { o.getFolioCorto(), o.getProveedor().getNombre(),
						String.format("$%.2f", o.getTotal()), o.getTipoPago().name(), o.getEstado().name(),
						String.valueOf(o.getId()) });
			}
		}
	}

	public void refrescarCuentas() {
		if (modeloCuentas == null)
			return;
		modeloCuentas.setRowCount(0);
		for (CuentaPorPagar c : ctrl.getCuentas())
			modeloCuentas.addRow(new Object[] { c.getProveedor().getNombre(), c.getFolioOrden(),
					String.format("$%.2f", c.getMontoTotal()), String.format("$%.2f", c.getPagado()),
					String.format("$%.2f", c.getSaldo()), c.getEstado().name(), String.valueOf(c.getId()) });
	}

	private void abrirFormularioProveedor(Proveedor p) {
		ProveedorDialog dlg = new ProveedorDialog(SwingUtilities.getWindowAncestor(this), p, ctrl);
		dlg.setVisible(true);
		refrescarProveedores();
	}

	private void abrirNuevaOrden() {
		NuevaOrdenDialog dlg = new NuevaOrdenDialog(SwingUtilities.getWindowAncestor(this), ctrl);
		dlg.setVisible(true);
		refrescarOrdenes();
	}
}

File: Vista/ProveedorPanel.py
from PyQt6.QtWidgets import (QWidget, QVBoxLayout, QHBoxLayout, QLabel, QPushButton, 
                             QTabWidget, QTableWidget, QTableWidgetItem, QHeaderView, 
                             QInputDialog, QMessageBox, QDialog)
from PyQt6.QtCore import Qt
from PyQt6.QtGui import QFont, QColor
from Vista.Estilos import Estilos
from Modelo.OrdenCompra import Estado as EstadoOrden, TipoPago
from Modelo.CuentaPorPagar import Estado as EstadoCuenta

class ProveedorPanel(QWidget):
    def __init__(self, ctrl, usuario):
        super().__init__()
        self.ctrl = ctrl
        self.usuario = usuario
        self.construir()

    def construir(self):
        self.setStyleSheet(f"background-color: {Estilos.BG_CLARO.name()};")
        layout = QVBoxLayout(self)
        layout.setContentsMargins(24, 24, 24, 24)
        layout.setSpacing(16)

        header = QWidget()
        header.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()}; border: 1px solid {Estilos.BORDE.name()};")
        h_layout = QVBoxLayout(header)
        h_layout.setContentsMargins(24, 20, 24, 20)
        tit = QLabel("ABASTECIMIENTO  —  Proveedores y Cuentas por Pagar")
        tit.setFont(Estilos.FUENTE_TITULO)
        tit.setStyleSheet("border: none;")
        h_layout.addWidget(tit)
        layout.addWidget(header)

        self.tabs = QTabWidget()
        self.tabs.setFont(Estilos.FUENTE_BOLD)
        self.tabs.addTab(self.construirTabProveedores(), "Directorio")
        self.tabs.addTab(self.construirTabOrdenes(), "Ordenes")
        self.tabs.addTab(self.construirTabCuentas(), "Deudas")
        
        layout.addWidget(self.tabs)

    def construirTabProveedores(self):
        p = QWidget()
        p.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()};")
        layout = QVBoxLayout(p)
        layout.setContentsMargins(12, 12, 12, 12)
        layout.setSpacing(8)

        toolbar = QHBoxLayout()
        toolbar.addStretch()
        btnNuevo = Estilos.botonPrimario("+ Alta Proveedor")
        btnNuevo.setFixedSize(160, 40)
        btnNuevo.clicked.connect(lambda: self.abrirFormularioProveedor(None))
        toolbar.addWidget(btnNuevo)
        layout.addLayout(toolbar)

        self.tablaProveedores = QTableWidget(0, 6)
        self.tablaProveedores.setHorizontalHeaderLabels(["ID", "Empresa", "Contacto", "Telefono", "Email", "Direccion"])
        self.tablaProveedores.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
        self.tablaProveedores.setColumnWidth(0, 60)
        self.tablaProveedores.setEditTriggers(QTableWidget.EditTrigger.NoEditTriggers)
        self.tablaProveedores.setSelectionBehavior(QTableWidget.SelectionBehavior.SelectRows)
        self.tablaProveedores.itemDoubleClicked.connect(self.onProveedorDoubleClick)
        
        layout.addWidget(self.tablaProveedores)
        return p

    def onProveedorDoubleClick(self, item):
        row = item.row()
        id_prov = int(self.tablaProveedores.item(row, 0).text())
        prov = next((x for x in self.ctrl.getProveedores() if x.getId() == id_prov), None)
        if prov:
            self.abrirFormularioProveedor(prov)

    def construirTabOrdenes(self):
        p = QWidget()
        p.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()};")
        layout = QVBoxLayout(p)
        layout.setContentsMargins(12, 12, 12, 12)
        layout.setSpacing(8)

        toolbar = QHBoxLayout()
        toolbar.addStretch()
        
        btnBitacora = Estilos.botonSecundario("Bitacora de Entradas")
        btnBitacora.setFixedSize(190, 40)
        btnBitacora.clicked.connect(self.mostrarBitacoraEntradas)
        toolbar.addWidget(btnBitacora)
        
        btnNueva = Estilos.botonExito("+ Generar Pedido")
        btnNueva.setFixedSize(170, 40)
        btnNueva.clicked.connect(self.abrirNuevaOrden)
        toolbar.addWidget(btnNueva)
        
        layout.addLayout(toolbar)

        self.tablaOrdenes = QTableWidget(0, 6)
        self.tablaOrdenes.setHorizontalHeaderLabels(["Folio", "Proveedor", "Total", "Pago", "Estado", "Accion"])
        self.tablaOrdenes.horizontalHeader().setSectionResizeMode(1, QHeaderView.ResizeMode.Stretch)
        self.tablaOrdenes.setColumnWidth(0, 100)
        self.tablaOrdenes.setColumnWidth(3, 100)
        self.tablaOrdenes.setColumnWidth(4, 110)
        self.tablaOrdenes.setColumnWidth(5, 110)
        
        layout.addWidget(self.tablaOrdenes)
        return p

    def recibir_orden_accion(self, oid):
        orden = next((o for o in self.ctrl.getOrdenes() if o.getId() == oid), None)
        if orden and orden.getEstado() == EstadoOrden.RECIBIDO:
            QMessageBox.information(self.window(), "Aviso", "Esta orden ya fue recibida.")
            return
        self.ctrl.recibirOrden(oid)
        self.refrescarOrdenes()
        self.refrescarCuentas()

    def construirTabCuentas(self):
        p = QWidget()
        p.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()};")
        layout = QVBoxLayout(p)
        layout.setContentsMargins(12, 12, 12, 12)
        layout.setSpacing(8)

        subtit = QLabel("  Saldos pendientes con acreedores")
        subtit.setFont(Estilos.FUENTE_SMALL)
        subtit.setStyleSheet(f"color: {Estilos.TEXTO_TENUE.name()}; border: none;")
        layout.addWidget(subtit)

        self.tablaCuentas = QTableWidget(0, 7)
        self.tablaCuentas.setHorizontalHeaderLabels(["Proveedor", "Orden", "Total", "Pagado", "Saldo", "Estado", "Accion"])
        self.tablaCuentas.horizontalHeader().setSectionResizeMode(0, QHeaderView.ResizeMode.Stretch)
        self.tablaCuentas.setColumnWidth(5, 100)
        self.tablaCuentas.setColumnWidth(6, 120)
        
        layout.addWidget(self.tablaCuentas)
        return p

    def liquidar_cuenta_accion(self, cid):
        cuenta = next((c for c in self.ctrl.getCuentas() if c.getId() == cid), None)
        if cuenta and cuenta.getSaldo() > 0:
            monto, ok = QInputDialog.getDouble(self.window(), "Liquidar", f"Monto a liquidar (saldo: ${cuenta.getSaldo():.2f}):", cuenta.getSaldo(), 0.01, cuenta.getSaldo(), 2)
            if ok:
                self.ctrl.liquidarCuenta(cid, monto)

    def refrescarProveedores(self):
        self.tablaProveedores.setRowCount(0)
        for i, p in enumerate(self.ctrl.getProveedores()):
            self.tablaProveedores.insertRow(i)
            self.tablaProveedores.setItem(i, 0, QTableWidgetItem(str(p.getId())))
            self.tablaProveedores.setItem(i, 1, QTableWidgetItem(p.getNombre()))
            self.tablaProveedores.setItem(i, 2, QTableWidgetItem(p.getContacto()))
            self.tablaProveedores.setItem(i, 3, QTableWidgetItem(p.getTelefono()))
            self.tablaProveedores.setItem(i, 4, QTableWidgetItem(p.getEmail()))
            self.tablaProveedores.setItem(i, 5, QTableWidgetItem(p.getDireccion()))

    def refrescarOrdenes(self):
        self.tablaOrdenes.setRowCount(0)
        for o in self.ctrl.getOrdenes():
            if o.getEstado() == EstadoOrden.PENDIENTE:
                row = self.tablaOrdenes.rowCount()
                self.tablaOrdenes.insertRow(row)
                self.tablaOrdenes.setItem(row, 0, QTableWidgetItem(o.getFolioCorto() if hasattr(o, 'getFolioCorto') else str(o.getId())))
                self.tablaOrdenes.setItem(row, 1, QTableWidgetItem(o.getProveedor().getNombre()))
                self.tablaOrdenes.setItem(row, 2, QTableWidgetItem(f"${o.getTotal():.2f}"))
                self.tablaOrdenes.setItem(row, 3, QTableWidgetItem(o.getTipoPago().name))
                self.tablaOrdenes.setItem(row, 4, QTableWidgetItem(o.getEstado().name))
                
                btnRecibir = QPushButton("Recibir")
                btnRecibir.setStyleSheet(f"background-color: {Estilos.EMERALD_CLARO.name()}; color: {Estilos.EMERALD.name()}; border: none;")
                btnRecibir.clicked.connect(lambda checked, oid=o.getId(): self.recibir_orden_accion(oid))
                self.tablaOrdenes.setCellWidget(row, 5, btnRecibir)

    def refrescarCuentas(self):
        self.tablaCuentas.setRowCount(0)
        for c in self.ctrl.getCuentas():
            row = self.tablaCuentas.rowCount()
            self.tablaCuentas.insertRow(row)
            self.tablaCuentas.setItem(row, 0, QTableWidgetItem(c.getProveedor().getNombre()))
            self.tablaCuentas.setItem(row, 1, QTableWidgetItem(c.getFolioOrden() if hasattr(c, 'getFolioOrden') else str(c.getOrdenId())))
            self.tablaCuentas.setItem(row, 2, QTableWidgetItem(f"${c.getMontoTotal():.2f}"))
            self.tablaCuentas.setItem(row, 3, QTableWidgetItem(f"${c.getPagado():.2f}"))
            self.tablaCuentas.setItem(row, 4, QTableWidgetItem(f"${c.getSaldo():.2f}"))
            self.tablaCuentas.setItem(row, 5, QTableWidgetItem(c.getEstado().name))
            
            btnLiquidar = QPushButton("Liquidar")
            btnLiquidar.setStyleSheet(f"background-color: {Estilos.INDIGO_CLARO.name()}; color: {Estilos.INDIGO.name()}; border: none;")
            btnLiquidar.clicked.connect(lambda checked, cid=c.getId(): self.liquidar_cuenta_accion(cid))
            self.tablaCuentas.setCellWidget(row, 6, btnLiquidar)

    def mostrarBitacoraEntradas(self):
        ordenes = [o for o in self.ctrl.getOrdenes() if o.getEstado() == EstadoOrden.RECIBIDO]
        dlg = QDialog(self.window())
        dlg.setWindowTitle("Bitacora de Entradas de Mercancia")
        dlg.setFixedSize(860, 520)
        layout = QVBoxLayout(dlg)
        layout.setContentsMargins(20, 20, 20, 20)
        
        tit = QLabel("BITACORA DE ENTRADAS DE MERCANCIA")
        tit.setFont(Estilos.FUENTE_TITULO)
        layout.addWidget(tit)
        
        sub = QLabel(f"{len(ordenes)} recepciones registradas")
        layout.addWidget(sub)
        
        tabla = QTableWidget(0, 6)
        tabla.setHorizontalHeaderLabels(["Folio", "Proveedor", "Fecha Recepcion", "Tipo Pago", "Total", "Productos Recibidos"])
        tabla.horizontalHeader().setSectionResizeMode(5, QHeaderView.ResizeMode.Stretch)
        tabla.setEditTriggers(QTableWidget.EditTrigger.NoEditTriggers)
        
        for o in ordenes:
            row = tabla.rowCount()
            tabla.insertRow(row)
            tabla.setItem(row, 0, QTableWidgetItem(f"#{o.getId()}"))
            tabla.setItem(row, 1, QTableWidgetItem(o.getProveedor().getNombre()))
            fecha_str = o.getFecha().strftime("%d/%m/%Y %H:%M") if o.getFecha() else ""
            tabla.setItem(row, 2, QTableWidgetItem(fecha_str))
            tabla.setItem(row, 3, QTableWidgetItem(o.getTipoPago().name))
            tabla.setItem(row, 4, QTableWidgetItem(f"${o.getTotal():.2f}"))
            prods = ", ".join([f"{i.getProducto().getNombre()} x{int(i.getCantidadSolicitada())}" for i in o.getItems()])
            tabla.setItem(row, 5, QTableWidgetItem(prods))
            
        if not ordenes:
            tabla.insertRow(0)
            tabla.setItem(0, 1, QTableWidgetItem("Sin recepciones registradas"))
            
        layout.addWidget(tabla)
        btnCerrar = QPushButton("Cerrar")
        btnCerrar.clicked.connect(dlg.accept)
        layout.addWidget(btnCerrar, alignment=Qt.AlignmentFlag.AlignCenter)
        dlg.exec()

    def abrirFormularioProveedor(self, p):
        from Vista.ProveedorDialog import ProveedorDialog
        dlg = ProveedorDialog(self.window(), p, self.ctrl)
        dlg.exec()
        self.refrescarProveedores()

    def abrirNuevaOrden(self):
        from Vista.NuevaOrdenDialog import NuevaOrdenDialog
        dlg = NuevaOrdenDialog(self.window(), self.ctrl)
        dlg.exec()
        self.refrescarOrdenes()
